resources/2024-10-dapr-bindings.md

```diff
@@ -0,0 +1,276 @@
+
+```mermaid
+graph LR
+    Client -->|Request| Engine
+    Engine -->|Deploy| Kubernetes
+    Application -->|Invokes| Dapr_Sidecar
+    Dapr_Sidecar -->|1 - Resolves| Dapr_Binding
+    Dapr_Sidecar --> |2 - call| Service_X
+
+    subgraph Foreign
+        Service_X
+    end
+
+    subgraph Pod
+        Application
+        Dapr_Sidecar
+    end
+
+    subgraph Kubernetes
+        Pod
+        Dapr_Binding
+    end
+```	
+
+#### Sequence Diagram
+
+( No changes in existing interactions)
+
+### Detailed Design
+
+This design will require adding a new type to the Dapr RP. This includes:
+- Adding Application.Dapr/Bindings to the TypeSpec frontend.
+- Adding the corresponding DaprBinding type to the internal representation (DaprRP)
+- Allowing a new type of Dapr component (binding) to be emitted in the backend Kubernetes cluster.
+
+#### Advantages (of each option considered)
+
+The main advantage of this implementation approach is that it is purely additive and will be non-breaking for existing users.
+
+#### Disadvantages (of each option considered)
+
+The main disadvantage of this approach is that it will lead to some code duplication.
+
+#### Proposed Option
+
+This option is the most straightforward and efficient in terms of development time. Since the Dapr Building Block implementation is not yet complete, it would be better to have a simple implementation that can be improved later, even if it introduces some code duplication. Additionally, future implementations of User-Defined Types may lead to a refactor of the entire Dapr RP.
+
+### API design (if applicable)
+
+Aside from the new type, the API will remain identical to the other Dapr Building Blocks.
+
+**typespec/Applications.Dapr/bindings.tsp**
+```TypeSpec
+model DaprBindingResource
+  is TrackedResourceRequired<DaprBindingProperties, "DaprBindings"> {
+  @doc("Binding name")
+  @key("bindingName")
+  @path
+  @segment("bindings")
+  name: ResourceNameString;
+}
+
+@doc("Dapr binding portable resource properties")
+model DaprBindingProperties {
+  ...EnvironmentScopedResource;
+  ...DaprResourceProperties;
+
+  @doc("A collection of references to resources associated with the binding")
+  resources?: ResourceReference[];
+
+  ...RecipeBaseProperties;
+}
+```
+
+### Implementation Details
+
+As for every Dapr Building Block, the implementation will need to create in the Dapr RP:
+- A versioned/unversioned api type converter
+- A dedicated processor for the Dapr Binding type
+
+This implementation will also have some side-effects outside of the Dapr RP:
+- Updates to the allowed resource type in the Portable Resource Renderer
+- Updates to the `getResourceDataByID` function in the Core RP to take the new type into account 
+- Updates to the `ResourceTypesList` variable in the CLI (cli/clients/managements.go) to be able to list the new type
+
+### Error Handling
+
+No new error handling is required. The error handling will remain the same as for other Dapr Building Blocks.
+
```

Comment:
Can we describe briefly how the system will handle errors when communicating with the external systems in case  they are down or unresponsive?

Reply:
I've added some information on that topic, but it may be out of scope as it describes how the Dapr sidecar reacts.